### ISSUE TITLE
Open cl environment

### ISSUE THREAD

**Issue by 18521613** [thinkoco/c5soc_opencl#26]
Hi @thinkoco 
Could you tell me how can create De1-SoC linux image with OpenCL driver and init_opencl_.sh file? I have spent a lot of time looking for any documents about it but i couldn't find anything!
Thanks you!

**Comment by thinkoco** [comment]
Hi，

A:
 1.You can downlaod the c5soc_opencl_lxde_fpga_reconfigurable_20201027.img
2. do as follows，and choose de1-soc  in step 2
https://github.com/thinkoco/c5soc_opencl#run-default-opencl-application

Now, the sdcard is for de1-Soc with de1-Soc sharedonly-vga BSP

B:
For building your own sdcard,here are some guides.

https://github.com/thinkoco/c5soc_opencl/blob/master/documents/LinuxAndDrivers.md#how-to-update-opencl-driver

https://github.com/thinkoco/c5soc_opencl/blob/master/documents/HowToBuildSDImage.md

**Comment by 18521613** [comment]
hi @thinkoco,
I want to build my sdcard so i have tried to do your B option with socfpga-opencl_4.1.22 branch on your git. But when i use _cp config_opencl .config_ , the system informs that no file such a config_opencl or .config file. I have clicked on show hidden file but there is no file named config_opencl or .config

**Comment by thinkoco** [comment]
@18521613
Hi, the socfpga-3.18  is a stable branch for opencl. 
Also, linux 4.9 and 4.19 has been tested with modified opencl driver.Here are some guides.
https://github.com/thinkoco/c5soc_opencl_rte/blob/master/documents/How_to_do.md

**Comment by 18521613** [comment]
@thinkoco I'm following your guides: https://github.com/thinkoco/c5soc_opencl_rte/blob/master/documents/How_to_do.md with linux 4.9. But when I'd used command _make zImage_, i got this error 
![image](https://user-images.githubusercontent.com/80325868/111865791-a8cee680-899b-11eb-972e-16d188d89899.png)

**Comment by thinkoco** [comment]
hi @18521613 
have you check out the socfpga-4.9.78-aoc barnch ?
```
git clone --depth 3 --branch socfpga-4.9.78-aocl https://github.com/thinkoco/linux-socfpga.git
export ARCH=arm
export CROSS_COMPILE=arm-linux-gnueabihf-
export LOADADDR=0x8000
make c5socl_defconfig
make zImage -j4
make modules
make socfpga_cyclone5_de10_nano.dtb
```

**Comment by 18521613** [comment]
> hi @18521613
> have you check out the socfpga-4.9.78-aoc barnch ?
> 
> ```
> git clone --depth 3 --branch socfpga-4.9.78-aocl https://github.com/thinkoco/linux-socfpga.git
> export ARCH=arm
> export CROSS_COMPILE=arm-linux-gnueabihf-
> export LOADADDR=0x8000
> make c5socl_defconfig
> make zImage -j4
> make modules
> make socfpga_cyclone5_de10_nano.dtb
> ```

Yes, the last picture i sent to you that is socfpga-4.9.78-aocl branch and when i used make _zImage -j4_ i got the same error. And i have a question that do i have to do _Update linux kernel_ part in your guide: https://github.com/thinkoco/c5soc_opencl_rte/blob/master/documents/How_to_do.md  with socfpga-4.9.78-aocl branch
And  one more when i try with linux 4.19 i also got this error:
![image](https://user-images.githubusercontent.com/80325868/111899885-e6964280-8a61-11eb-9730-732491b12d06.png)

**Comment by thinkoco** [comment]
@18521613 
for linux 4.9, no need to update files after checkout the ` socfpga-4.9.78-aocl` branch
```
git clone --depth 3 --branch socfpga-4.9.78-aocl https://github.com/thinkoco/linux-socfpga.git
export ARCH=arm
export CROSS_COMPILE=arm-linux-gnueabihf-
export LOADADDR=0x8000
make c5socl_defconfig
make zImage -j4
make modules
make socfpga_cyclone5_de10_nano.dtb
```
for linux 4.19, update the driver files  , makefile and kconfig

**Comment by 18521613** [comment]
Hi @thinkoco,
I have built my sd card image with opencl driver successfully. But when i try to execute hello_world example of BSP(Board Support Package) for Altera SDK OpenCL 18.1 from terasic (http://download.terasic.com/downloads/cd-rom/de1-soc/linux_BSP/OPENCL18.1/DE1-SoC_OpenCL_BSP.zip) i got this error:
![image](https://user-images.githubusercontent.com/80325868/114065871-4fd0df00-98c5-11eb-9cc1-f1b4102f92cd.png)
Do you know the way to solve this error? Thanks you!

**Comment by thinkoco** [comment]
@18521613 
maybe, the sdcard is unpluged  when data copying. just fix it by  `fsck`

**Comment by tungdt113** [comment]
> @18521613
> maybe, the sdcard is unpluged when data copying. just fix it by `fsck`

Thanks for your help, I have solved that issue. But i got another issue :(. I have run `init_opencl_18.1_4.19.sh` but when i try to execute hello_world.aocx the system inform that _Couldn't open /dev/ac10 device! Did you load the driver?_
![image](https://user-images.githubusercontent.com/64633842/114837952-6d8ad080-9dfe-11eb-86d2-e5221d5064a5.png)

And when i run `ls /dev` there is no ac10 here! Do you know anything about this? 
![image](https://user-images.githubusercontent.com/64633842/114837809-4cc27b00-9dfe-11eb-9b46-47878809eba7.png)

Thank you!

**Comment by thinkoco** [comment]
@tungdt113 
 Hi, the opencl driver should be rebuild when running with a new zImage. 
1. dowload the runtime enviroment files here. 
https://github.com/thinkoco/c5soc_opencl_rte/blob/master/linux-4.19.x/opencl_rte/aocl-rte-18.1.0-625.arm32.4.19.tar.xz
2. [here](https://github.com/thinkoco/c5soc_opencl/blob/master/documents/LinuxAndDrivers.md#how-to-update-opencl-driver) is the guide for buiding opencl driver
```
export ARCH=arm
export CROSS_COMPILE=arm-linux-gnueabihf-
cd  xxxxxx/aocl-rte-18.1.0-625.arm32.4.19/board/c5soc/arm32/driver
make KDIR=xxxxxx/linux-socfpga
```

**Comment by 18521613** [comment]
Hi @thinkoco,
I have built opencl driver and copy it to /home/root in partiotion 2 of sdcard, but it not working..

**Comment by thinkoco** [comment]
Hi @18521613 
you may change the path in the `init_opencl_18.1_4.19 .sh` for your case , or copy it to `$AOCL_BOARD_PACKAGE_ROOT/arm32/driver/`
```
#!/bin/bash
export INTELFPGAOCLSDKROOT=/root/aocl-rte-18.1.0-625.arm32.4.19
export AOCL_BOARD_PACKAGE_ROOT=$INTELFPGAOCLSDKROOT/board/c5soc
export PATH=$INTELFPGAOCLSDKROOT/bin:$PATH
export LD_LIBRARY_PATH=$INTELFPGAOCLSDKROOT/host/arm32/lib:$AOCL_BOARD_PACKAGE_ROOT/arm32/lib:$LD_LIBRARY_PATH
insmod $AOCL_BOARD_PACKAGE_ROOT/arm32/driver/aclsoc_drv.ko
#export CL_CONTEXT_COMPILER_MODE_INTELFPGA=3
```

**Comment by tungdt113** [comment]
@thinkoco,
I have load  aclsoc_drv.ko successfully but there is still no ac10 device here
![image](https://user-images.githubusercontent.com/64633842/115332428-fe2c2c80-a1c1-11eb-994a-06e9ff203178.png)

**Comment by tungdt113** [comment]
Hi @thinkoco,
i have tried to build OpenCL driver like this guide: https://github.com/thinkoco/c5soc_opencl/blob/master/documents/LinuxAndDrivers.md. I have downloaded Intel FPGA Runtime Environment for OpenCL Linux Cyclone V SoC TGZ 17.1 on Altera Download Website. And I build it for kernel 3.18 like your image: _c5soc_opencl_lxde_fpga_reconfigurable_20201027.img_. But when I try to test my OpenCL driver on your image. I got this error like this:
![image](https://user-images.githubusercontent.com/64633842/120219378-5ba5a580-c265-11eb-86a9-f9d3044f4fa9.png)
P/s: I've download kernel3.18 in your git: https://github.com/thinkoco/linux-socfpga/tree/socfpga-3.18

**Comment by tungdt113** [comment]
Hi @thinkoco 
I've successfully generate "/dev/acl0" when source init_opencl.sh file. But when i try to execute the hello example from terasic. I got this error.
![image](https://user-images.githubusercontent.com/64633842/122325018-07344280-cf54-11eb-8320-929af64de175.png)
i've checked /sys/class/fpga_bridge folder there is no folder have name like them
![image](https://user-images.githubusercontent.com/64633842/122325207-5a0dfa00-cf54-11eb-8a8f-ee2011e6ceef.png)
But when i've checked /sys/class/fpga_bridge in your linux image, it has the above folders. So could you tell me how to generate them? 
Thank you!

**Comment by 18521613** [comment]
> @thinkoco I'm following your guides: https://github.com/thinkoco/c5soc_opencl_rte/blob/master/documents/How_to_do.md with linux 4.9. But when I'd used command _make zImage_, i got this error
> ![image](https://user-images.githubusercontent.com/80325868/111865791-a8cee680-899b-11eb-972e-16d188d89899.png)

Hi @thinkoc,
 I have built another image with linuxs-socfpga 4.9.78-ltsi so my zImage doesn't change fpga bridges name for aclsoc driver. I hope you could help me solve this error.
Thank you!!

**Comment by thinkoco** [comment]
@18521613  
the error occurs when using gcc 5.0 and above, it has been fixed. you can checkout the lastest code on branch `socfpga-4.9.78-aocl`

**Comment by tungdt113** [comment]
@thinkoco,
I've rebuild zimage on `socfpga-4.9.78-aocl` branch, it builds successfully. But when boot this zImage and socfpga.dtb for de1_soc that i take from in your image : `c5soc_opencl_lxde_fpga_reconfigurable_20201027.img` , it seemly don't add /dev/fpga0 char driver and file operations.
![image](https://user-images.githubusercontent.com/64633842/122631960-b65e4e80-d0f9-11eb-935d-660909b58713.png)
And i have a question that if i can use `socfpga_cyclone5_de1soc.dts` from `socfpga-3.18` branch to build `socfpga_cyclone5_de1soc.dtb` for `socfpga-4.9.78-aocl`?? When i use this dtb to boot, it stop booting kernel like this:
`socfpga-4.9.78-aocl`
![image](https://user-images.githubusercontent.com/64633842/122631982-dc83ee80-d0f9-11eb-927d-ea1d1631f297.png)

**Comment by tungdt113** [comment]
@thinkoco
i have a question that if i can use `socfpga_cyclone5_de1soc.dts` from `socfpga-3.18` branch to build `socfpga_cyclone5_de1soc.dtb` for `socfpga-4.9.78-aocl` ??

**Comment by tungdt113** [comment]
> @18521613
> the error occurs when using gcc 5.0 and above, it has been fixed. you can checkout the lastest code on branch `socfpga-4.9.78-aocl`

@thinkoco,
I've rebuild with your update, it builds successfully. But when i boot this `zImage` with `socfpga.dtb` for de1soc that i take from your image: `c5soc_opencl_lxde_fpga_reconfigurable_20201027.img`. It seemly don't add /dev/fpga0 char driver and file operations. And i saw you change something in `c5soc_opencl_rte/blob/master/socfpga-4.9.78-ltsi/include/linux/fpga/fpga-mgr.h`, but i did'n see any change in` fpga-mgr.h` of `socfpga-4.9.78-aocl` branch
![image](https://user-images.githubusercontent.com/64633842/122635324-68534600-d10d-11eb-9e0c-9077731c5add.png)
And there are some warnings when i build zImage with your update. Is there anything problem with this??
![image](https://user-images.githubusercontent.com/64633842/122635392-def04380-d10d-11eb-8b19-5a1fad35fa72.png)

**Comment by thinkoco** [comment]
@tungdt113 
yes, it’s [here](https://github.com/thinkoco/linux-socfpga/commit/cb24d83e6ef8d742d8a396248c83ececff5fff89)
